### DISCUSSION TITLE
Track safe usage of 'unsafe' type parameters

### DISCUSSION THREAD

**Discussion by KotlinIsland** [KotlinIsland/basedmypy#224] [Ideas]
Here's an idea, is it possible to track safe usage of type parameters in potentially unsafe positions?

```py
class A(Generic[T_co]):
    def foo(self, t: T_co):
        print(t)
```
This code is perfectly safe (although a subtype could change that) and ideally should be allowed.

It would be super interesting to investigate the bounds of what's possible in terms of "safe variance"

**Comment by DetachHead** [comment]
is this related to the `SafeVariance` idea

**Reply by KotlinIsland** [reply]
Actually it's related to `UnsafeVariance` #209

**Reply by KotlinIsland** [reply]
Thinking about this a bit more, it might be a more refined form of `SafeVariance`, where the parameter is basically `object`.
